core/src/main/java/io/grpc/internal/RetriableStream.java

```diff
@@ -89,6 +92,10 @@
    */
   private final AtomicBoolean noMoreTransparentRetry = new AtomicBoolean();
 
+  /** Keeps track of all substreams for informative purposes. */
+  private final ConcurrentLinkedQueue<Substream> allSubstreams =
```

Comment:
I'm concerned about how long we'll keep these streams alive. We could potentially be retaining an entire transport for "no reason."

I think we should add a new `Collection<Status>` or `Collection<Status.Code>` to `State` for the completed streams. The stream's Status can be easily passed to `State.substreamClosed(Substream)`, and so when the stream is removed from `drainedSubstreams` its results will be present in the other list. There's no need for the other stream data for completed streams (like address) since it isn't normally provided.

Reply:
I have updated this PR according to your suggestion. Now it includes the status codes from all closed substreams. @dapengzhang0, according to javadoc on the `drainedSubstreams` field, it doesn't seem to always include unfinished substreams. Please advice on how to proceed.

Reply:
Per offline discussion, I have updated the code and added comments.